Sources/CacheAPI-smf.php

```diff
@@ -0,0 +1,169 @@
+	 * @param string $ttl Overrides the default TTL.
+	 * @return mixed The result from the cache, if there is no data or it is invalid, we return null.
+	 */
+	public function getData($key, $ttl = null)
+	{
+		global $cachedir;
+
+		$key = $this->prefix . strtr($key, ':/', '-_');
+
+		// SMF Data returns $value and $expired.  $expired has a unix timestamp of when this expires.
+		if (file_exists($cachedir . '/data_' . $key . '.php') && filesize($cachedir . '/data_' . $key . '.php') > 10)
+		{
+			// Work around Zend's opcode caching (PHP 5.5+), they would cache older files for a couple of seconds
+			// causing newer files to take effect a while later.
+			if (function_exists('opcache_invalidate'))
+				opcache_invalidate($cachedir . '/data_' . $key . '.php', true);
+
+			if (function_exists('apc_delete_file'))
+				@apc_delete_file($cachedir . '/data_' . $key . '.php');
+
+			// php will cache file_exists et all, we can't 100% depend on its results so proceed with caution
+			@include($cachedir . '/data_' . $key . '.php');
+			if (!empty($expired) && isset($value))
+			{
+				@unlink($cachedir . '/data_' . $key . '.php');
+				unset($value);
+			}
+		}
+
+		return !empty($value) ? $value : null;
+	}
+
+	/**
+	 * Saves to data the cache.
+	 *
+	 * @access public
+	 * @param string $key The key to use, the prefix is applied to the key name.
+	 * @param mixed $value The data we wish to save.
+	 * @param string $ttl Overrides the default TTL.
+	 * @return bool Whether or not we could save this to the cache.
+	 */
+	public function putData($key, $value, $ttl = null)
+	{
+		global $cachedir;
+
+		$key = $this->prefix . strtr($key, ':/', '-_');
+
+		// Work around Zend's opcode caching (PHP 5.5+), they would cache older files for a couple of seconds
+		// causing newer files to take effect a while later.
+		if (function_exists('opcache_invalidate'))
+			opcache_invalidate($cachedir . '/data_' . $key . '.php', true);
+
+		if (function_exists('apc_delete_file'))
+			@apc_delete_file($cachedir . '/data_' . $key . '.php');
+
+		// Otherwise custom cache?
+		if ($value === null)
+			@unlink($cachedir . '/data_' . $key . '.php');
+		else
+		{
+			$cache_data = '<' . '?' . 'php if (!defined(\'SMF\')) die; if (' . (time() + $ttl) . ' < time()) $expired = true; else{$expired = false; $value = \'' . addcslashes($value, '\\\'') . '\';}' . '?' . '>';
+
+			// Write out the cache file, check that the cache write was successful; all the data must be written
+			// If it fails due to low diskspace, or other, remove the cache file
+			$fileSize = file_put_contents($cachedir . '/data_' . $key . '.php', $cache_data, LOCK_EX);
+			if ($fileSize !== strlen($cache_data))
+			{
+				@unlink($cachedir . '/data_' . $key . '.php');
+				return false;
+			}
+			else
+				return true;
+		}
+	}
+
+	/**
+	 * Closes connections to the cache method.
+	 *
+	 * @access public
+	 * @return bool Whether or not we could close connections.
+	 */
+	public function quit()
```

Comment:
these functions that don't do anything should be moved to the base class, less duplication.